notebook/nbextensions.py

```diff
@@ -285,132 +663,451 @@ class InstallNBExtensionApp(JupyterApp):
+        if logger:
+            logger.warn("Out of date: %s" % dest)
+        return True
+    if logger:
+        logger.info("Up to date: %s" % dest)
+    return False
+
+
+def _maybe_copy(src, dest, logger=None):
+    """Copy a file if it needs updating.
+
+    Parameters
+    ----------
+
+    src : string
+        A path that should exist from which to copy a file
+    src : string
+        A path that might exist to which to copy a file
+    logger : Jupyter logger [optional]
+        Logger instance to use
+    """
+    if _should_copy(src, dest, logger=logger):
+        if logger:
+            logger.info("Copying: %s -> %s" % (src, dest))
+        shutil.copy2(src, dest)
+
+
+def _safe_is_tarfile(path):
+    """Safe version of is_tarfile, return False on IOError.
+
+    Returns whether the file exists and is a tarfile.
+
+    Parameters
+    ----------
+
+    path : string
+        A path that might not exist and or be a tarfile
+    """
+    try:
+        return tarfile.is_tarfile(path)
+    except IOError:
+        return False
+
+
+def _get_nbextension_dir(user=False, sys_prefix=False, prefix=None, nbextensions_dir=None):
+    """Return the nbextension directory specified
+
+    Parameters
+    ----------
+
+    user : bool [default: False]
+        Get the user's .jupyter/nbextensions directory
+    sys_prefix : bool [default: False]
+        Get sys.prefix, i.e. ~/.envs/my-env/share/jupyter/nbextensions
+    prefix : str [optional]
+        Get custom prefix
+    nbextensions_dir : str [optional]
+        Get what you put in
+    """
+    if sum(map(bool, [user, prefix, nbextensions_dir, sys_prefix])) > 1:
+        raise ArgumentConflict("cannot specify more than one of user, sys_prefix, prefix, or nbextensions_dir")
+    if user:
+        nbext = pjoin(jupyter_data_dir(), u'nbextensions')
+    elif sys_prefix:
+        nbext = pjoin(ENV_JUPYTER_PATH[0], u'nbextensions')
+    elif prefix:
+        nbext = pjoin(prefix, 'share', 'jupyter', 'nbextensions')
+    elif nbextensions_dir:
+        nbext = nbextensions_dir
+    else:
+        nbext = pjoin(SYSTEM_JUPYTER_PATH[0], 'nbextensions')
+    return nbext
+
+
+def _nbextension_dirs():
+    """The possible locations of nbextensions.
+
+    Returns a list of known base extension locations
+    """
+    return [
+        pjoin(jupyter_data_dir(), u'nbextensions'),
+        pjoin(ENV_JUPYTER_PATH[0], u'nbextensions'),
+        pjoin(SYSTEM_JUPYTER_PATH[0], 'nbextensions')
+    ]
+
+
+def _get_config_dir(user=False, sys_prefix=False):
+    """Get the location of config files for the current context
+
+    Returns the string to the enviornment
+
+    Parameters
+    ----------
+
+    user : bool [default: False]
+        Get the user's .jupyter config directory
+    sys_prefix : bool [default: False]
+        Get sys.prefix, i.e. ~/.envs/my-env/etc/jupyter
+    """
+    if user and sys_prefix:
+        raise ArgumentConflict("Cannot specify more than one of user or sys_prefix")
```

Comment:
Yeah, see the issue. Changed the defaults for `user` on enable.

It should probably be default `None`, and handle it with a switch if nothing else is set. This is too bad, as the docstrings are nice. I'll get on it this evening!